test/lit/passes/heap2local.wast

```diff
@@ -0,0 +1,1527 @@
+  ;; CHECK-NEXT:   (drop
+  ;; CHECK-NEXT:    (local.get $ref)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (local.get $2)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $safe-to-drop-multiflow (result f64)
+    (local $ref (ref null $struct.A))
+    (local.set $ref
+      (struct.new_default_with_rtt $struct.A
+        (rtt.canon $struct.A)
+      )
+    )
+    ;; An extra drop + multiple flows through things do not stop us.
+    (drop
+      (block (result (ref null $struct.A))
+        (block (result (ref null $struct.A))
+          (loop (result (ref null $struct.A))
+            (local.get $ref)
+          )
+        )
+      )
+    )
+    (struct.get $struct.A 1
+      (local.get $ref)
+    )
+  )
+
+  ;; CHECK:      (func $escape-after-multiflow (result f64)
+  ;; CHECK-NEXT:  (local $ref (ref null $struct.A))
+  ;; CHECK-NEXT:  (local.set $ref
+  ;; CHECK-NEXT:   (struct.new_default_with_rtt $struct.A
+  ;; CHECK-NEXT:    (rtt.canon $struct.A)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (call $send-ref
+  ;; CHECK-NEXT:   (block (result (ref null $struct.A))
+  ;; CHECK-NEXT:    (block (result (ref null $struct.A))
+  ;; CHECK-NEXT:     (local.get $ref)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (struct.get $struct.A 1
+  ;; CHECK-NEXT:   (local.get $ref)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $escape-after-multiflow (result f64)
+    (local $ref (ref null $struct.A))
+    (local.set $ref
+      (struct.new_default_with_rtt $struct.A
+        (rtt.canon $struct.A)
+      )
+    )
+    ;; An escape after multiple flows.
+    (call $send-ref
+      (block (result (ref null $struct.A))
+        (block (result (ref null $struct.A))
+          (loop (result (ref null $struct.A))
+            (local.get $ref)
+          )
+        )
+      )
+    )
+    (struct.get $struct.A 1
+      (local.get $ref)
+    )
+  )
+
+  ;; CHECK:      (func $non-exclusive-set (result f64)
+  ;; CHECK-NEXT:  (local $ref (ref null $struct.A))
+  ;; CHECK-NEXT:  (local.set $ref
+  ;; CHECK-NEXT:   (select (result (ref $struct.A))
+  ;; CHECK-NEXT:    (struct.new_default_with_rtt $struct.A
+  ;; CHECK-NEXT:     (rtt.canon $struct.A)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (struct.new_default_with_rtt $struct.A
+  ;; CHECK-NEXT:     (rtt.canon $struct.A)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (i32.const 1)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (struct.get $struct.A 1
+  ;; CHECK-NEXT:   (local.get $ref)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $non-exclusive-set (result f64)
+    (local $ref (ref null $struct.A))
+    ;; A set that receives two different allocations, and so we should not try
+    ;; to optimize it.
+    (local.set $ref
+      (select
+        (struct.new_default_with_rtt $struct.A
+          (rtt.canon $struct.A)
+        )
+        (struct.new_default_with_rtt $struct.A
+          (rtt.canon $struct.A)
+        )
+        (i32.const 1)
```

Comment:
Interesting point, yeah, they really are a tuple, logically...

In addition to what you said, we also don't have a way to set a tuple field, do we? Just to create a new tuple. That would be less readable even if we can optimize it out later. Also I guess we don't want to depend on the multivalue feature when using GC.

Reply:
We don't currently have a way to set a tuple field, but it would be fairly easy to add and would definitely improve readability if we wanted to do an optimization like this.

> Also I guess we don't want to depend on the multivalue feature when using GC.

I don't know about that; I would be very surprised to see an engine support GC but not multivalue.